### PR TITLE
Map AvroSchema to Arrow (#4886)

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Part of #4886

# Rationale for this change
 

The logic to map an avro schema to arrow is not entirely straightforward, this PR adds the logic necessary to perform this transformation.

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: arrow-avro/src/lib.rs [hunk 1]
```diff
@@ -27,6 +27,8 @@ mod schema;
 
 mod compression;
 
+mod codec;
```
Comment (tustvold) [inline arrow-avro/src/lib.rs:30]: This is all kept crate-private to allow us to revisit this design as things evolve

File: arrow-avro/src/schema.rs [hunk 2]
```diff
@@ -94,8 +94,6 @@ pub enum Schema<'a> {
 #[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
 #[serde(tag = "type", rename_all = "camelCase")]
 pub enum ComplexType<'a> {
-    #[serde(borrow)]
-    Union(Vec<Schema<'a>>),
```
Comment (tustvold) [inline arrow-avro/src/schema.rs:98]: This was a mistake, Unions are encoded at a higher-level

File: arrow-avro/src/codec.rs [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+use std::collections::HashMap;
+use std::sync::Arc;
+
+/// Avro types are not nullable, with nullability instead encoded as a union
+/// where one of the variants is the null type.
+///
+/// To accommodate this we special case two-variant unions where one of the
+/// variants is the null type, and use this to derive arrow's notion of nullability
+#[derive(Debug, Copy, Clone)]
+enum Nulls {
```
Comment (tustvold) [inline arrow-avro/src/codec.rs:32]: This whole setup is rather annoying, but necessary to provide a reasonable API

File: arrow-avro/src/codec.rs [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+                name,
+                namespace,
+                resolver,
+            )?;
+            let meta = Arc::make_mut(&mut field.meta);
+
+            // https://avro.apache.org/docs/1.11.1/specification/#logical-types
+            match (t.attributes.logical_type, &mut meta.codec) {
+                (Some("decimal"), c @ Codec::Fixed(_)) => {
+                    return Err(ArrowError::NotYetImplemented(format!(
```
Comment (tustvold) [inline arrow-avro/src/codec.rs:265]: I don't intend to support decimals in a first version of this crate

File: arrow-avro/src/codec.rs [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+        let (namespace, name) = name
+            .rsplit_once('.')
+            .unwrap_or_else(|| (namespace.unwrap_or(""), name));
+
+        self.map
+            .get(&(namespace, name))
+            .ok_or_else(|| ArrowError::ParseError(format!("Failed to resolve {namespace}.{name}")))
+            .cloned()
+    }
+}
```
Comment (crepererum) [inline arrow-avro/src/codec.rs:161]: This could use some docstrings. For once: what's a namespace and why is the top-level namespace `item`?

**Comment by Samrose-Ahmed** [comment]
Looks good to me overall.

Btw does it even matter which position the null union is in (0,1), I guess you're keeping it to map correctly to the input.

**Review by crepererum — APPROVED** [review]

**Pull request merged**